### PR TITLE
Refactor: Migrate to Typescript/React Functional Component

### PR DESCRIPTION
**Pull request by erickriva**

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,22 +1,20 @@
 {
-  "extends": ["eslint:recommended", "plugin:react/recommended", "prettier"],
-  "parser": "@babel/eslint-parser",
-  "plugins": ["react"],
-  "rules": {
-    "react/jsx-uses-react": "error",
-    "react/jsx-uses-vars": "error",
-    "array-bracket-spacing": 0,
-    "no-trailing-spaces": 1,
-    "no-tabs": 1
```
Comment (sylvaindubus) [inline .eslintrc.json:10]: Please re-enable these three rules.

File: demo/index.html [hunk 2]
```diff
@@ -6,16 +6,12 @@
     <meta name="viewport" content="width=device-width, initial-scale=1" />
     <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1" />
     <link rel="preconnect" href="https://fonts.gstatic.com" />
-    <link rel="icon" type="image/png" sizes="32x32" href="static/favicon-32x32.png" />
-    <link rel="icon" type="image/png" sizes="16x16" href="static/favicon-16x16.png" />
```
Comment (sylvaindubus) [inline demo/index.html:10]: Why did you remove these two links?

File: demo/package.json [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
+    "start": "parcel index.html",
+    "build": "parcel build index.html",
     "publish": "npm run build && gh-pages -d dist",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "prestart": "npm link .."
+  },
+  "dependencies": {
+    "react": "^18.1.0",
+    "react-dom": "^18.1.0"
```
Comment (sylvaindubus) [inline demo/package.json:12]: I know this is a debatable point but I prefer build-time dependencies to be considered as `devDependencies`.

File: package.json [hunk 4]
```diff
@@ -17,35 +17,37 @@
...
-  "main": "dist/index.js",
+  "main": "./dist/cjs/index.js",
+  "module": "./dist/esm/index.js",
+  "types": "./dist/esm/index.d.ts",
   "scripts": {
-    "build": "babel src/index.jsx -d dist",
+    "build": "yarn build:esm && yarn build:cjs",
+    "build:esm": "tsc",
+    "build:cjs": "tsc --module commonjs --outDir dist/cjs",
+    "build-watch": "yarn build:esm -w && yarn build:cjs -w",
```
Comment (sylvaindubus) [inline package.json:27]: Even if I also like yarn, I would prefer to stay with npm for simplicity purposes.

File: demo/src/components/App/index.tsx [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import './App.css'
+
+function App() {
+  const prismaZoom = React.useRef<PrismaZoomRef>(null)
+
+  const [zoom, setZoom] = React.useState(1)
+  const [allowZoom, setAllowZoom] = React.useState(true)
+  const [allowPan, setAllowPan] = React.useState(true)
+
+  const onZoomChange = (zoom) => {
```
Comment (sylvaindubus) [inline demo/src/components/App/index.tsx:15]: The type of this parameter is missing (and same for `onDoubleClickOnCard`)

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,22 @@
+{
+  // Change this to match your project
+  "include": ["src/index.tsx", "src/types.ts"],
+  "compilerOptions": {
+    "outDir": "dist/esm",
+    "module": "esnext",
+    "target": "es5",
+    "lib": ["es6", "dom", "es2016", "es2017"],
```
Comment (sylvaindubus) [inline tsconfig.json:8]: Could you replace `es2016` and `es2017` with `esnext`?

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    "moduleResolution": "node",
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "esModuleInterop": true,
+    "noImplicitReturns": true,
+    "noImplicitThis": true,
+    "noImplicitAny": true,
+    "strictNullChecks": true,
+    "suppressImplicitAnyIndexErrors": true,
+    "allowSyntheticDefaultImports": true
```
Comment (sylvaindubus) [inline tsconfig.json:20]: I don't like these two :)

File: tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{
+  // Change this to match your project
```
Comment (sylvaindubus) [inline tsconfig.json:2]: We can remove this one once we are okay with this conf

File: demo/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (sylvaindubus) [inline demo/tsconfig.json:1]: Apply or discuss the same changes as in the main conf file

File: src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,559 @@
+import React from 'react'
```
Comment (sylvaindubus) [inline src/index.tsx:1]: I don't have too much to say about this migration, it's really great! I just have a preference for decomposing the React variable to get a bit more concise code (e.g `forwardRef` instead of `React.forwardRef`)

**Review by sylvaindubus — COMMENTED** [review]
Thank you for this great job, we are really going in the right direction with this lib!
Please feel free to give me your feedback on these comments.

File: demo/index.html (hunk 2 above)
Comment (erickriva) [reply demo/index.html:10]: I had removed because of some errors, but I'm putting it back again on its place.

File: src/index.tsx (hunk 10 above)
Comment (erickriva) [reply src/index.tsx:1]: I prefer the React.* notation, so I can differentiate React and project code, but is only personal taste. I'll be decomposing for next commit.

File: demo/tsconfig.json (hunk 9 above)
Comment (erickriva) [reply demo/tsconfig.json:1]: What do you think will be better?
To copy same configs from **main** to **demo**, so it will become clearer?
Or demo `tsconfig.json` must extend the **main** one?

Comment (sylvaindubus) [reply demo/tsconfig.json:1]: I think **demo** can extend the **main** one
